### ISSUE TITLE
Question: Work on texlab latex formatter

### ISSUE THREAD

**Issue by ModProg** [latex-lsp/texlab#497]
This is more a question than an issue, but discussions aren't enabled in this repository.

In the options documentation, it says that the `texlab` formatter for latex is not yet implemented.

Are there any concrete plans on doing so?

**Comment by pfoerster** [comment]
> In the options documentation, it says that the texlab formatter for latex is not yet implemented.

That's correct. At the moment, `texlab` uses `latexindent` under the hood. However, I am bit unhappy with the fact that `latexindent` is very slow and is awkward to configure. Ideally, we would have our own formatter but it is a lot of effort to create a fast (and reliable) formatter. Besides that, we need some changes in the parser (for example we treat "&" as ordinary text in the parser).

> Are there any concrete plans on doing so?

Not at the moment. In the meantime, I would be very happy to include a contributed formatter (either here in this repository or as a separate tool/library).

> but discussions aren't enabled in this repository.

They should be enabled now 👍 .

**Comment by ModProg** [comment]
Makes sense, I just didn't find anything on this. Thanks for the answer.

**Comment by alxiong** [comment]
Upvote for releasing a fast latex & bibtex formatter as a separate executable. 

(Ideally I'd like to configure my emacs/lsp to use this formatter by default, and use pre-commit hooks to enforce formatting on other developers' machine where they don't use emacs)

**Comment by clason** [comment]
But if it's an external tool (of which there already are several, like https://github.com/nschloe/blacktex), why should it be shoehorned into an independent LSP server instead of going through standard editor hooks for formatting (like Vim's `'formatprg'`, which I'm sure Emacs has an analogue of)? 

> use pre-commit hooks to enforce formatting on other developers' machine where they don't use emacs

And, if I may say so, that would be the quickest method of getting me to stop collaborating with you on a paper ;) If you have _that_ restrictive ideas about writing LaTeX, you get to write all of it by yourself...

**Comment by alxiong** [comment]
thanks for the suggestion @clason, I'm quite a noob when it comes to Tex tooling, 

may I ask what's the benefit of an independent formatter in latex-lsp over hooking up with an existing external formatter? 

> if you have that restrictive ideas about writing LaTeX, you get to write all of it by yourself...

haha, good to know the objection, in my mind, the formatting rules can be agreeable among all authors right? it's more about being consistency than being autocratic on style, would you agree? 
(I guess I won't complain about a project with `pyproject.toml` or `.prettierjs.json` or `rustfmt.toml`, if anything I'd hope they explicitly states their idiosyncrasies and enforce them automatically)

**Comment by clason** [comment]
> may I ask what's the benefit of an independent formatter in latex-lsp over hooking up with an existing external formatter?

There is none -- an external formatter should be independent, hooking it up with a language server is just extra work and complexity for no gain.

**Comment by clason** [comment]
> in my mind, the formatting rules can be agreeable among all authors right

Oh, you sweet summer child...

>  it's more about being consistent than being autocratic on style, would you agree?

In the end, people will only care about the compiled PDF -- and if it's for publication, the publisher will take the TeX files and put them through their own thresher, anyway. I'd concentrate on the contents rather than the code style...

**Comment by ModProg** [comment]
> In the end, people will only care about the compiled PDF -- and if it's for publication, the publisher will take the TeX files and put them through their own thresher, anyway. I'd concentrate on the contents rather than the code style...

You could say that about any code formatting, still many people use formatters to help them write more readable code. And depending on your workflow it could also help with e.g. merging changes.

> There is none -- an external formatter should be independent, hooking it up with a language server is just extra work and complexity for no gain.

Well depending on how you use them and what editor you use, it might help if the LSP knows how to use the formatter, but I think it does make sense to implement it separately and just use it as a library in here and make it available stand alone as well.

**Comment by clason** [comment]
> You could say that about any code formatting, still many people use formatters to help them write more readable code. And depending on your workflow it could also help with e.g. merging changes.

The difference is that a codebase needs to be maintained long-term, while a paper or book is hopefully done at some point.

> Well depending on how you use them and what editor you use, it might help if the LSP knows how to use the formatter, but I think it does make sense to implement it separately and just use it as a library in here and make it available stand alone as well.

I don't see how the LSP should know how to use the formatter if it's external -- it's just passing the command through, so you might as well bypass the middleman. It's different if the internal tooling of the server generates an intermediate representation that can be leveraged for code formatting, but that doesn't sound like it's the case here.

**Comment by nolanking90** [comment]
@pfoerster Do you have any specific thoughts on implementing this with the current state of texlab? Specifically, can you think of any things that would need to be changed in the parser/syntax tree?

I've been prototyping some stuff for this. I have a visitor that walks the tree and naively tries to reconstruct the document. I haven't thought deeply about what the formatting algorithm or formatting rules should be.  Right now, it just handles a few cases like begin and end environments. Here's a small example:

```latex
\documentclass{article}
\usepackage{geometry}
\begin{document}
\section{First Section}\label{asec:outer} % 1 Section
\begin{equation}\label{eq:myEQ}
1 + 1 = 2
\end{equation}
\subsection{Subsection}{First Subsection} % 1.1 Subsection
\section{Another Section} % 2 Section'
some stuff.
\subsection{Dummy} % 2.1 Dummy
some more stuff.
\subsection*{Subsection'}\label{sec:inner} % Subsection'
\end{document}
```
gets formatted to 
```latex
\documentclass{article}
\usepackage{geometry}

\begin{document}

\section{First Section}\label{asec:outer}
	\begin{equation}\label{eq:myEQ}
		1 + 1 = 2
	\end{equation}

	\subsection{Subsection}{First Subsection}

\section{Another Section}
	some stuff.

	\subsection{Dummy}
		some more stuff.

	\subsection{Subsection'}\label{sec:inner}
\end{document}
```

1. Here I had to change some things in the parser for the + inside the equation to be preserved. (I don't remember if I was getting 1 1 = 2 or just the first 1 before). You can see things like inline comments seem to be missing from the tree too (or you would have to return to some ancestor of the current node to find the comment or to even know that it was supposed to be there), and things like unnumbered variants of commands (e.g. \section*) don't keep the * in their command names. I'm really not sure if this is a problem with the tree, something intentional, or if I'm just not understanding something about how to use Rowan (the Rowan docs are a little terse). It seems like it would be helpful to have more information stored in each node, like having a SectionNode struct that contains a field like unnumbered:bool that can be checked while producing the formatted output (I'm assuming node is a trait that SectionNode could implement and not an explicit struct that Rowan defines). 

2. I have been using the petgraph crate to output graphviz DOT to the logfile, and copy/pasting that into another program, so I can look at the actual syntax tree while I'm troubleshooting. Does Rowan have a way to do something like this builtin? This is kind of a pain to do, so if there is a better way to inspect the tree please let me know.

3. I also think I should write tests before I start working on specific formatting rules. Do you think tests should be tex to tex, or should they take a root node of some already parsed tex and produce tex? For the tests in the parser, did you write the expected output by hand, or is there a way to generate the text representation of a tree using Rowan? I mean things like 
```rust
r#"
      ROOT@0..48
        PREAMBLE@0..48
          TEXT@0..4
            WORD@0..3 "Foo"
            WHITESPACE@3..4 "\n"
          BLOCK_COMMENT@4..22
            COMMAND_NAME@4..12 "\\iffalse"
            WHITESPACE@12..13 "\n"
            WORD@13..18 "Test1"
            WHITESPACE@18..19 "\n"
            COMMAND_NAME@19..22 "\\fi"
          WHITESPACE@22..23 "\n"
"#
```

Let me know if you have any other thoughts or suggestions here. 
I will clean up my local repo soon and make a draft PR for this so you can see the state of things if you want. 
I don't expect to have anything ready to merge until the spring (or possibly the summer depending on my work).

**Comment by pfoerster** [comment]
@nolanking90

> Specifically, can you think of any things that would need to be changed in the parser/syntax tree?

I am currently thinking about splitting the parsing into two stages:
1. Do the rough parsing (curly braces, environments, sections, commands) first and keep this as the syntax tree.
2. Do the specific parsing (special commands like includes, references, citations, glossary entries) at a later stage on-demand.

Compared to the current approach, where everything is parsed at once, this has the following advantages:
- It is easier to handle constructs where definitions in other documents influence the semantics of the documents (see #537)
- Less prone to reporting false-positives since only the rough parsing could be used to report syntax errors
- Make the parsing itself a whole lot simpler without hard-coding that many special cases

I think this approach could also make it easier when writing a formatter. You can gradually add more complexity to the formatting rules without breaking the remaining rules or altering the parser.

> You can see things like inline comments seem to be missing from the tree too

Rowan syntax trees are lossless so it's always possible to fully reconstruct the original text even if the document has faulty syntax. When walking the syntax tree, you always need to go down to the token level, where the original text is stored. The trivia nodes (comments, whitespace) are attached directly to the syntax nodes. If possible, the parser prefers to move the trivia nodes as far up as possible in the tree so that the ranges of nodes like `\section` are not unnecessarily large (this is especially relevant for the folding ranges).

>  Do you think tests should be tex to tex, or should they take a root node of some already parsed tex and produce tex?

I think, it's easiest if the tests work on tex to tex. Using parsed trees makes it hard to maintain if the parser and thus the syntax tree changes for a given document.

> For the tests in the parser, did you write the expected output by hand, or is there a way to generate the text representation of a tree using Rowan?

For the `expect!` tests, you can use the `UPDATE_EXPECT=1` environment variable to generate the write to actual output to the expected output inside the test files. For `rust-analyzer`, there is also an extra command that does this for you selectively for a single test or module of tests.

https://github.com/rust-analyzer/expect-test
https://docs.rs/expect-test/latest/expect_test/

> Does Rowan have a way to do something like this builtin?

Usually, I run the debug formatter on the parse tree (`format!("{:?}", root)`. For a graphical ouput, I don't think there is anything better at the moment than using `petgraph` and `graphviz`.

**Comment by nolanking90** [comment]
Thanks! You're right, I was not checking the tokens. Here's the syntax tree in a cute graph now that I'm walking it correctly.

![image](https://github.com/user-attachments/assets/a17927af-8b52-4512-8c83-3047147a7963)

I'll post another update when I make some more progress on the actual formatter.

**Comment by nolanking90** [comment]
@pfoerster I made a new crate for the formatter but I can't seem to get tests to run on it. Rust analyzer says the file is untracked and suggests adding mod tests to my lib.rs, but it doesn't look like any of the other crates do this. cargo test is telling me there are no tests run or ignored. Thoughts?

https://github.com/nolanking90/texlab/tree/feature/formatter/crates/formatter

**Comment by nolanking90** [comment]
I have a prototype of the formatter working on my fork [github.com/nolanking90/texlab](github.com/nolanking90/texlab)

It needs some more polishing before I put together a PR (e.g. remove some unwraps). 
I also have not pulled from the main repo in a while, so it's a little behind, I'm sure. 
However, I'm not sure that I can make much more progress on it without some people trying it out and giving some feedback.

**Comment by jolars** [comment]
If you're interested in switching to/adding a different formatter (other than latexindent), perhaps you'd be interested in looking at my project [Badness](https://badness.dev)? Its formatter is published separately as a crate at <https://crates.io/crates/badness-formatter>, so I think including it as an opt-in in texlab would be relatively straightforward. It does feed off a different parser and CST shape though, so might be best to be a bit careful.
